lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,371 @@
+        Args:
+            relation: The relation for which to set the vault url.
+            address: The address of the vault server.
+
+        """
+        if not self.charm.unit.is_leader():
+            logger.warning(
+                "Attempting to set the vault url without being the leader. Ignoring the request."
+            )
+            return
+        relation.data[self.charm.app].update({"address": address})
+
+    def set_credentials_secret_id(self, relation: ops.Relation, secret_id: str) -> None:
+        """Set the credentials secret id in the relation databag.
+
+        Args:
+            relation: The relation for which to set the credentials secret id.
+            secret_id: The secret id of the Juju secret which stores the credentials for authenticating with the Vault server.
+        """
+        if relation.app is None:
+            raise ValueError("The `relation.app` is not set")
+        relation.data[self.charm.app].update({"credentials_secret_id": secret_id})
+
+    def set_ca_certificate(self, relation: ops.Relation, ca_certificate: str) -> None:
+        """Set the ca_certificate on the relation.
+
+        Args:
+            relation: The relation for which to set the ca_certificate.
+            ca_certificate: The CA certificate to use when validating the Vault server's certificate.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        if not relation:
+            logger.warning("Relation is None")
+            return
+        if not relation.active:
+            logger.warning("Relation is not active")
+            return
+        relation.data[self.charm.app]["ca_certificate"] = ca_certificate
+
+
+def is_provider_data_valid(data: RelationDataContent) -> bool:
+    """Return whether the provider data is valid.
+
+    This uses the pydantic schema to validate the data.
+
+    Args:
+        data: The data to validate.
+
+    Returns:
+        True if the data is valid, False otherwise.
+    """
+    try:
+        ProviderSchema(app=VaultAutounsealProviderSchema(**data))
+        return True
+    except ValidationError as e:
+        logger.warning("Invalid data: %s", e)
+        return False
+
+
+class VaultAutounsealRequires(ops.Object):
+    """Manages the vault-autounseal relation from the requirer side."""
+
+    on: VaultAutounsealRequireEvents = VaultAutounsealRequireEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.relation = relation_name
+
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        if event.app is None:
+            logger.warning("No remote application yet")
+            return
+        if is_provider_data_valid(event.relation.data[event.app]):
+            self.on.vault_autounseal_details_ready.emit(event.relation)
```

Comment:
Why pass the whole relation instead of the actual content that is needed? What would the requirer do with the relation object?

Reply:
Well, we would need at least the ID from the relation to properly configure autounseal. We could do that and also return the databag contents directly, but we actually won't use this event other than to fire `_configure()` anyway. 

If we wanted it implement a standalone handler in the future (for some reason, if it becomes needed), I figured we could get everything we needed from the relation (since there are already getters that take the relation, which we need for `_configure`)

Oh, and you guessed it, I also copied it from vault_kv :laughing: .

Happy to change if you think it would look cleaner one way or the other.

Reply:
fixed in https://github.com/canonical/vault-k8s-operator/pull/346/commits/0651467cd3a75065ba0f04391194ceef96844526

Reply:
> Well, we would need at least the ID from the relation to properly configure autounseal.

Why does the requirer need the ID? It needs the vault url and credentials right? Why not pass those instead? I think we will likely limit this relation to `1` from the requirer side removing any ambiguity. This way we should be able to run run `self.vault_autounseal.get_vault_url()`?

In any case, if we don't plan on using the event handler at all, I don't think it makes sense to keep it.

Reply:
The path to the transit key is based on the relation id, but I suppose we could (and probably should) just add a method to get that.

To properly configure it will need

- vault url
- approle id
- approle secret id
- ca cert
- key  name (based on relation id)
- mount path (constant)

Reply:
https://github.com/canonical/vault-k8s-operator/pull/346/commits/c9675c04e4090f53cf9f9fd62a6b9463875ca6a4

Except the part about adding a way to get the key path instead of inferring it, that will come in a later commit.

Reply:
Transit key path method added in https://github.com/canonical/vault-k8s-operator/pull/346/commits/14fd1ad51eb98cd6682bdc5af57dfe29e709cec1

Reply:
Why did [7e9e247](https://github.com/canonical/vault-k8s-operator/pull/346/commits/7e9e247c566b543488e6c7676467f709e3652ecb) remove the changes introduced in [c9675c0](https://github.com/canonical/vault-k8s-operator/commit/c9675c04e4090f53cf9f9fd62a6b9463875ca6a4) of passing the information to the event?

Reply:
Because I did it late at night :laughing: 

Oops, I must I have done something wrong when switching between branches. Fixed in https://github.com/canonical/vault-k8s-operator/pull/346/commits/c23518ce430bfb6d5566c1e20e0d1acc4247c1a8